docs/src/main/asciidoc/qute.adoc

```diff
@@ -0,0 +1,271 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Qute Templating Engine
+
+include::./attributes.adoc[]
+
+Qute is a templating engine designed specifically to meet the Quarkus needs.
+The usage of reflection is minimized to reduce the size of native images.
+The API combines both the imperative and the non-blocking reactive style of coding.
+In the development mode, all files located in `META-INF/resources/templates` are watched for changes and modifications are immediately visible.
+Furthermore, we try to detect most of the template problems at build time.
+In this guide, you will learn how to easily render templates in your application.
+
+[NOTE]
+====
+This extension is considered `preview`.
+API or configuration properties might change as the extension matures.
+Feedback is welcome on our https://groups.google.com/d/forum/quarkus-dev[mailing list] or as issues in our https://github.com/quarkusio/quarkus/issues[GitHub issue tracker].
+====
+
+== Hello World with JAX-RS
+
+If you want to use Qute in your JAX-RS application, you need to add the `quarkus-qute-resteasy` extension first.
+In your `pom.xml` file, add:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-qute-resteasy</artifactId>
+</dependency>
+----
+
+We'll start with a very simple template:
+
+.hello.txt
+----
+Hello {name}! <1>
+----
+<1> `{name}` is a value expression that is evaluated when the template is rendered.
+
+NOTE: By default, all files located in the `META-INF/resources/templates` directory and its subdirectories are registered as templates. Templates are validated during startup and watched for changes in the development mode.
+
+Now let's inject the "compiled" template in the resource class.
+
+.HelloResource.java
+[source,java]
+----
+package org.acme.quarkus.sample;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+
+import io.quarkus.qute.Rendering;
+import io.quarkus.qute.Template;
+
+@Path("hello")
+public class HelloResource {
+
+    @Inject
+    Template hello; <1>
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public Rendering get(@QueryParam("name") String name) {
```

Comment:
I find `Rendering` to be a very confusing name. Perhaps `TemplateInstance` or `TemplateRendering`, but it's not really a "rendering instance". It is a template instance, with data. Perhaps the actual `Template` should be `TemplateFactory` and this `Rendering` should be `Template`. After all, in Java we never use `Foo` and `FooInstance`, we would call them `FooBuilder/Factory` and `Foo` for the instances.

Reply:
It's not a template instance. It's "template instance + data + some config". So the best name would be probably `RenderingConfig` or `RenderingContext`. -1 for `TemplateFactory`. We can discuss the name later though.

Reply:
Not really useful, but in the reports module in Seam 3 we had `ReportDefinition` and `Report`: https://github.com/seam/reports/tree/develop/api/src/main/java/org/jboss/seam/reports

Reply:
Well, to me, "template instance" can hold data. Like "Map" holds data.

Reply:
Hm, `TemplateDefinition` sounds good. @FroMage WDYT?

@gastaldi `ReportDefinition` javadoc: _"A report definition object (normally called a "template")"_. ;-)

Reply:
@mkouba yeah, some Report engines call it that way :)

Reply:
I'm totally fine with `TemplateDefinition` for the field, and `Template` for the object returned by the REST method.

Reply:
Hm, I think I will try the `Template` -> `TemplateInstance` first. It sounds more natural to me. I'll update the PR and we can continue the discussion afterwards ;-).